zapd/templates/amdevice_list.html

```diff
@@ -4,4 +4,7 @@
 <li>
     <a href="multiple_devices" title="Show With Multiple Devices">Show With Multiple Devices</a>
 </li>
+<li>
+    <a href="" title="Update Address List">Update</a>
```

Comment:
href is not filled

Reply:
trying to figure out how I could get the "Update" button to run:
```
    @classmethod
    def initial_wallet_addresses(cls, session):
        # update txs
        limit = 100
        issuer_addr = issuer_address(app.config["NODE_ADDRESS"], app.config["ASSET_ID"])
        oldest_txid = None
        txs = []
        addrs = {}
        while True:
            have_tx = False
            txs = blockchain_transactions(app.config["NODE_ADDRESS"], issuer_addr, limit, oldest_txid)
            for tx in txs:
                oldest_txid = tx["id"]
                if tx["type"] == 4 and tx["assetId"] == app.config["ASSET_ID"]:
                    if 'sender' in tx:
                        sender = tx['sender']
                        addrs[sender] = 1
                    if 'recipient' in tx:
                        recipient = tx['recipient']
                        addrs[recipient] = 1
                if have_tx or len(txs) < limit:
                    break
        for key, value in addrs:
            wallet = AMWAllet(key)
            session.add(wallet)
            session.add(AMDevice(wallet, 'n/a', 'n/a', 'n/a', 'n/a', 'n/a', 'n/a'))
        session.commit()
```
Which is under the AMWallet(db.Model).

At the moment, I have "@expose('update_address_list')" going to a webpage which has :
```
    <a href="" title="Update Address List">Update</a>
```

So its currently not doing anything. The question is how to get from the webpage to the AMWallet(db.Model)

Reply:
href still is empty (https://www.w3schools.com/tags/att_a_href.asp)

Reply:
Placing:
```
<li>
    <a href="/update_list" title="Update Address List">Update</a>
</li>
```

Which means i need to put something in the web.py. The question then would be what do i put in the web.py:
```
@app.route("/update_list", methods=["POST"])
def update_list():
     this is what im trying to figure out. Need something in here that would use that section in the @classmodel for performing the query to loop through the blockchain
    return "ok"
```

Reply:
no we are using the flask admin expose (https://flask-admin.readthedocs.io/en/latest/api/mod_base/#flask_admin.base.expose) decorator which sets up the route for us

you have:
`@expose("/update_address_list")`

so why dont you put:
`<a href="/update_address_list" ...`

Reply:
ive changed the naming i used local repo to match remote repo.

Still ahve the problem of what i need to put in the web.py:
```
@app.route("/update_list", methods=["POST"])
def update_list():
     ###this is what im trying to figure out. Need something in here that would use that section in the @classmodel for performing the query to loop through the blockchain and im not able to figure it out.
    return "ok"
```

Currently its returning "ok" on my screen. 

changing it to make it look like this:
```
@app.route("/update_address_list", methods=["GET"])
def update_address_list():
    AMWallet.initial_wallet_addresses(db.session)

    return 'ok'
```

And now getting an error:
```
[2020-06-12 01:17:13,288] ERROR in app: Exception on /update_address_list [GET]
Traceback (most recent call last):
  File "/root/git/zapd/zapd/venv/lib/python3.6/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/root/git/zapd/zapd/venv/lib/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/root/git/zapd/zapd/venv/lib/python3.6/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/root/git/zapd/zapd/venv/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/root/git/zapd/zapd/venv/lib/python3.6/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/root/git/zapd/zapd/venv/lib/python3.6/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/root/git/zapd/zapd/web.py", line 325, in update_address_list
    AMWallet.initial_wallet_addresses(db.session)
  File "/root/git/zapd/zapd/models.py", line 225, in initial_wallet_addresses
    issuer_addr = issuer_address(app.config["NODE_ADDRESS"], app.config["ASSET_ID"])
KeyError: 'NODE_ADDRESS'
```

So it looks like i can't use app.config["NODE_ADDRESS"] from the models.py

This is the @classmethod fromo the models.py:
```
    @classmethod
    def initial_wallet_addresses(cls, session):
        # update txs
        limit = 100
        issuer_addr = issuer_address(app.config["NODE_ADDRESS"], app.config["ASSET_ID"])
        oldest_txid = None
        txs = []
        addrs = {}
        while True:
            have_tx = False
            txs = blockchain_transactions(app.config["NODE_ADDRESS"], issuer_addr, limit, oldest_txid)
            for tx in txs:
                oldest_txid = tx["id"]
                if tx["type"] == 4 and tx["assetId"] == app.config["ASSET_ID"]:
                    if 'sender' in tx:
                        sender = tx['sender']
                        addrs[sender] = 1
                    if 'recipient' in tx:
                        recipient = tx['recipient']
                        addrs[recipient] = 1
                if have_tx or len(txs) < limit:
                    break
        for key, value in addrs:
            wallet = AMWAllet(key)
            session.add(wallet)
            session.add(AMDevice(wallet, 'n/a', 'n/a', 'n/a', 'n/a', 'n/a', 'n/a'))
        session.commit()
```
Which most likely going to fail for the other app.config["ASSET_ID"] as well.

Any ideas on how I could use app.config in the models.py?

Reply:
since app.config["NODE_ADDRESS"] and app.config["ASSET_ID"] is not in the script, im static placing it in the models.py

checking to see what else we need/testing.

Reply:
managed to get the following in the console:
```
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=3G6TvYnwmfQj2fhSKggUwHYjHGaQTLgSbX6p2ueQHnpd..
:: retrieved 1000 records
1000
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=J9RGNURR7PkghXdkVWu3oekfEfxcd87uFvtDLYrRFRfj..
:: retrieved 1000 records
1000
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=9BMiuGTDiCjzeXSV7fh8MRGaKugpeSN1zU6LPaTbU3AH..
:: retrieved 1000 records
1000
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=Hd2CDyAEJBcFKD4gqDEGcFDDTsXr6jnQx8AHcnC8cy8D..
:: retrieved 1000 records
1000
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=46DQCkg61PwhWsAkZmVM5tMSRYBeJgakKYCrfRSHw9Zq..
:: retrieved 1000 records
1000
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=8twKCvFHeK6L1heBaeSSQGFUrQcV6S5xTfV9iLdJLA34..
:: retrieved 836 records
1
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=8BnLfbbh8K5qsVuPus2DTCe3RNvJYiBJXwCxUj4t85u..
:: retrieved 835 records
1
```

Not sure if this a ratelimit of some sort? Or maybe that if the no. of records is less than the limit its starts requesting the exact same query with the next id in the last result.

Need to look at it further.

Reply:
so my one appears to run the blockchain_transaction and instead of looping in there, it would go out of the function and then go back in.

the question is why is it remaining in the loop and not breaking when the no. of transaction is below the limit.

need to find out why this is and force it break afterwards so i could look at writing the result to the db.

Reply:
so the loop is not breaking because I'm trying to break from inside the if statement. the moment I placed that in the same level as the for loop it breaks. Which is what I wanted. Checking to see if it would now write to the sqlite db. dropping the tables and then re-run the app.

Reply:
pulled the changes and it looks like its working:
![image](https://user-images.githubusercontent.com/47799652/84608156-2d6c7000-af05-11ea-940d-447abbf6de8c.png)

running the initialize again to see the result:
![image](https://user-images.githubusercontent.com/47799652/84608188-51c84c80-af05-11ea-921a-5667c39560ad.png)


Looks like its working